### ISSUE TITLE
Fatal interaction with the `request_store` gem and `RACK_ENV`

### ISSUE THREAD

**Issue by Earlopain** [Shopify/pitchfork#117]
Pardon the title but I don't really know how else to put it.

I use the [request_store](https://github.com/steveklabnik/request_store) gem in my app and after deploying to production I immediatly noticed quite the issue (left is latency, the pink increase on the right is time spent in pg):
![image](https://github.com/Shopify/pitchfork/assets/14981592/51eddefb-8254-4e24-8b20-3cb2f0132028)

I only deployed to one of three hosts so the spike for that one is even more bad than this average shows. I've narrowed it down to the following code pattern:

```rb
module DeferredPosts
  extend ActiveSupport::Concern

  def deferred_post_ids
    RequestStore[:deferred_post_ids] ||= Set.new
  end

  def deferred_posts
    Post.includes(:uploader).where(id: deferred_post_ids.to_a).find_each.reduce({}) do |post_hash, p|
      post_hash[p.id] = p.minimal_attributes
      post_hash
    end
  end
end
```

This is included in the `ApplicationHelper` and will, over the span of one request, accumulate some model ids to later load all at once. It's never manually cleared as this would normally happen automatically.

What I observed is that every request increases the the set further, degrading performance even more. Eventually the database was saturating the network link of the host.

I can only reproduce under these circumstances:
* Pitchfork is behind nginx
* `RACK_ENV` is set to production

I have created a repro at https://github.com/Earlopain/pitchfork_repro and hope this will work for you. To get it up an running:
* `docker compose up nginx` to start nginx. Will listen on port 3000 and proxy to port 8080.
* `RAILS_ENV=development RACK_ENV=production bundle exec pitchfork` to start pitchfork. Note that `RAILS_ENV=dev` is only for convenience and not part of the repro.

You can then access the site at http://localhost:3000/test/demo. Refreshing should yield something like this:

https://github.com/Shopify/pitchfork/assets/14981592/95741f8c-23aa-4401-8a6d-58da7d202359

For some requests it will return the same set and then be empty on the next one. I don't really see ryhme or reason to when that happens. http://localhost:8080/test/demo which goes to pitchfork directly functions as expected.

**Comment by casperisfine** [comment]
Alright, I'm able to repro, and as you mentioned, only with NGINX proxying, that is wild.

**Comment by Earlopain** [comment]
Tried something out, a friend of mine mentioned rack yesterday in relation to this and I did indeed do the upgrade from 2 -> 3 as well. In the repro, if you do `gem "rack", "~> 2.0"` is seems to behave as expected again

**Comment by casperisfine** [comment]
Yeah, I just found the issue. It's from `Rack::ContentLength`:

```ruby
module Rack

  # Sets the content-length header on responses that do not specify
  # a content-length or transfer-encoding header.  Note that this
  # does not fix responses that have an invalid content-length
  # header specified.
  class ContentLength
    include Rack::Utils

    def initialize(app)
      @app = app
    end

    def call(env)
      status, headers, body = response = @app.call(env)

      if !STATUS_WITH_NO_ENTITY_BODY.key?(status.to_i) &&
         !headers[CONTENT_LENGTH] &&
         !headers[TRANSFER_ENCODING] &&
         body.respond_to?(:to_ary)

        response[2] = body = body.to_ary # BUG: Discard Rack::BodyProxy
        headers[CONTENT_LENGTH] = body.sum(&:bytesize).to_s
      end

      response
    end
  end
end
```

I'll dig a bit more to see how this is suppsed to be handled.

**Comment by casperisfine** [comment]
Ok, so I asked for a release: https://github.com/rack/rack/pull/2062#issuecomment-2102291754

I'd suggest either sticking to `rack 2`, or running `rack` edge, or migrating to Rails current attributes which is resilient to this bug by design (state is also reset at request start).

**Comment by ioquatix** [comment]
This should be fixed in Rack v3.0.11, thanks for your effort diagnosing the issue.

**Comment by Earlopain** [comment]
Yeah, this should do it. I tested with the backport applied and that looked good. I'll get it to prod next week and I'm very confident this will work now. Though if not I'm going to report back somewhere.

Thanks everyone
